jujugui/static/gui/src/app/app.js

```diff
@@ -701,20 +689,44 @@ YUI.add('juju-gui', function(Y) {
       this.user.controller = { user, password, macaroons, external };
 
       controllerAPI.after('login', evt => {
+        const state = this.state;
+        const current = this.state.current;
         this.anonymousMode = false;
         if (evt.err) {
           this._renderLogin(evt.err);
           return;
         }
         this._renderUserMenu();
         console.log('successfully logged into controller');
+        // If state has a `next` property then that overrides all defaults.
+        const specialState = current.special;
+        const next = specialState && specialState.next;
+        // There should never be a `next` state if we aren't on login but
+        // adding the state login check here just to be sure.
+        if (state.current.root === 'login' && next) {
+          console.log('redirecting to "next" state', next);
+          const {error, state: newState} = state.generateState(next, false);
+          if (error === null) {
+            // The root at this point will be 'login' and because the `next`
+            // url may not explicitly define a new root path we have to set it
+            // to null to clear 'login' from the url.
+            if (!newState.root) {
+              newState.root = null;
+            }
+            newState.special = null;
+            this.maskVisibility(false);
+            state.changeState(newState);
+            return;
+          }
+          console.error('error redirecting to previous state', error);
```

Comment:
I usually agree with that, but in this case I felt that the execution flow of the code made more sense if it went the opposite here.